pennylane/transforms/decompositions/two_qubit_unitary.py

```diff
@@ -14,227 +14,202 @@
 
-def _extract_su2su2_prefactors(U, V):
-    r"""U, V are SU(4) matrices for which there exists A, B, C, D such that
-    (A \otimes B) V (C \otimes D) = U. The problem is to find A, B, C, D in SU(2)
-    in an analytic and fully differentiable manner.
 
-    This decomposition is possible when U and V are in the same double coset of
-    SU(4), meaning there exists G, H in SO(4) s.t. G (Edag V E) H = (Edag U
-    E). This is guaranteed here by how V was constructed using the
-    _select_rotation_angles method. Then, we can use the fact that E SO(4) Edag
-    gives us something in SU(2) x SU(2) to give A, B, C, D.
+def _decomposition_2_cnots(U, wires):
+    r"""If 2 CNOTs are required, we can write the circuit as
+     -U- = -A--X--RZ(d)--X--C-
+     -U- = -B--C--RX(p)--C--D-
+    We need to find the angles for the Z and X rotations such that the inner
+    part has the same spectrum as U, and then we can recover A, B, C, D.
     """
+    # Compute the rotation angles
+    u = math.dot(Edag, math.dot(U, E))
+    gammaU = math.dot(u, math.T(u))
+    evs, _ = math.linalg.eig(gammaU)
 
-    # A lot of the work here happens in the magic basis. Essentially, we
-    # don't look explicitly at some U = G V H, but rather at
-    #     E^\dagger U E = G E^\dagger V E H
-    # so that we can recover
-    #     U = (E G E^\dagger) V (E H E^\dagger) = (A \otimes B) V (C \otimes D).
-    # There is some math in the paper explaining how when we define U in this way,
-    # we can simultaneously diagonalize functions of U and V to ensure they are
-    # in the same coset and recover the decomposition.
-
-    u = math.dot(math.cast_like(Edag, V), math.dot(U, math.cast_like(E, V)))
-    v = math.dot(math.cast_like(Edag, V), math.dot(V, math.cast_like(E, V)))
-
-    uuT = math.dot(u, math.T(u))
-    vvT = math.dot(v, math.T(v))
-
-    # First, we find a matrix p (hopefully) in SO(4) s.t. p^T u u^T p is diagonal.
-    # Since uuT is complex and symmetric, both its real / imag parts share a set
-    # of real-valued eigenvectors.
-    if math.get_interface(u) == "tensorflow":
-        ev_p, p = math.linalg.eig(math.real(uuT))
+    # These choices are based on Proposition III.3 of
+    # https://arxiv.org/abs/quant-ph/0308045
+    # There is, however, a special case where the circuit has the form
+    # -U- = -A--C--X--C-
+    # -U- = -B--X--C--D-
+    #
+    # or some variant of this, where the two CNOTs are adjacent.
+    #
+    # What happens here is that the set of evs is -1, -1, 1, 1 and we can write
+    # -U- = -A--X-SZ-X--C-
+    # -U- = -B--C-SX-C--D-
+    # where SZ and SX are square roots of Z and X respectively. For some reason this
+    # case is not handled properly with the full algorithm, so we treat it separately.
+
+    if math.get_interface(u) == "torch":
+        # Torch's sort function returns both the sorted values and the new order
+        sorted_evs, _ = math.sort(math.real(evs))
```

Comment:
🤦 

Would you be able to make a very small GitHub issue regarding this? It would be nice to add a torch exception to `single_dispatch.py`, to make sure it conforms to TF and Numpy. Could be a nice first issue for someone, and will avoid future devs debugging the same thing